### ISSUE TITLE
Proposal - Make Better Use Of Typescript

### ISSUE THREAD

**Issue by Shepless** [fuse-box/fuse-box#930]
When I contribute I often find areas of the code that look completely unlike other areas and are sometimes hard to follow and read. In addition, I have found instances where we don't return consistent types from methods/functions. This means issues could sneak into the code base as a contributor may not return correct types, or assume incorrect types from an `any` type.

Therefore I'd like to propose the following:

- Stricter Type Checking (I.E. try to remove as many instances of `any`, explicit return Types, etc.)
- Introduce TSLint Rules

This will not only make the code base more uniform and easier for all to follow but should also aide in reducing issues through new features and also less regression issues.

**Comment by sveinburne** [comment]
I am quite confident with TS typings and would love to participate. @Shepless just ping me and spot a small area of the codebase you would like me to work on. If you are satisfied with my contributions, I could do more :-)

**Comment by Shepless** [comment]
@sveinburne awesome mate. Personally I'd start with all `Promise<T>` return types (which is a lot!) as fuse is relies very heavily on async and is probably the most vital area. Having said that, the ultimate aim (long term) should be no `any` types so any PR on any part you think can be improved is welcome!
